core/trino-spi/src/main/java/io/trino/spi/block/AbstractArrayBlock.java

```diff
@@ -116,22 +125,36 @@ public long getRegionSizeInBytes(int position, int length)
     }
 
     @Override
-    public long getPositionsSizeInBytes(boolean[] positions)
-    {
-        checkValidPositions(positions, getPositionCount());
-        boolean[] used = new boolean[getRawElementBlock().getPositionCount()];
-        int usedPositionCount = 0;
-        for (int i = 0; i < positions.length; ++i) {
-            if (positions[i]) {
-                usedPositionCount++;
-                int valueStart = getOffsets()[getOffsetBase() + i];
-                int valueEnd = getOffsets()[getOffsetBase() + i + 1];
-                for (int j = valueStart; j < valueEnd; ++j) {
-                    used[j] = true;
-                }
-            }
+    public final long getPositionsSizeInBytes(boolean[] positions, int selectedArrayPositions)
+    {
+        int positionCount = getPositionCount();
+        checkValidPositions(positions, positionCount);
+        if (selectedArrayPositions == 0) {
+            return 0;
+        }
+        if (selectedArrayPositions == positionCount) {
+            return getSizeInBytes();
+        }
+
+        Block rawElementBlock = getRawElementBlock();
+        OptionalInt fixedPerElementSizeInBytes = rawElementBlock.fixedSizeInBytesPerPosition();
+        int[] offsets = getOffsets();
+        int offsetBase = getOffsetBase();
+        long elementsSizeInBytes;
+
+        if (fixedPerElementSizeInBytes.isPresent()) {
+            elementsSizeInBytes = fixedPerElementSizeInBytes.getAsInt() * (long) countSelectedPositionsFromOffsets(positions, offsets, offsetBase);
+        }
+        else if (rawElementBlock instanceof RunLengthEncodedBlock) {
+            // RLE blocks don't have fixed size per position, but accept null for the positions array
+            elementsSizeInBytes = rawElementBlock.getPositionsSizeInBytes(null, countSelectedPositionsFromOffsets(positions, offsets, offsetBase));
```

Comment:
That's true. I left this in because I was planning on making `RunLengthEncodedBlock#getPositionsSizeInBytes` do something like `return selectedPositions > 0 ? value.getSizeInBytes() : 0;` but that broke a lot of tests and it didn't seem worth the headache- so I instead added tests that assert the opposite. It doesn't quite feel like the right for `AbstractArrayBlock` to special case the handling of `RunLengthEncodedBlock` by calling `RunLengthEncodedBlock#getSizeInBytes` instead, but it certainly would reduce the overhead involved.

I'm open to doing whatever you think makes the most sense here.